### PR TITLE
WIP - Refactor distgit module

### PR DESCRIPTION
**Pull request by thiagoalessio**
> **Work in Progress:** Already opening a PR to share where this is going, and allow the team to intervene earlier, comment or just be aware of the changes before I place the finished PR.
Thank you @sosiouxme for the suggestion.

> **Spoiler Alert:** The tests are _horrible_! There is a lot of repetition, patching & mocking going on.
Please keep in mind that this is necessary because the current production code is not test-friendly, as it depends heavily on state, such as metadata's "config" and "runtime".
We can considerably simplify the tests once we refactor the production code to make it more testable.
<br>Here is a draft of the direction I'd like to refactor:
<https://gist.github.com/thiagoalessio/c189b89f475d92fc2eb5347af1c65142>
Basically breaking down stuff to single-purposed functions and move logging outside function bodies.

## TLDR

Started covering `doozerlib/distgit.py` because is the worst file in the project, and the one that gets touched most frequently. So any improvement to this file means a higher chance of making our lives easier in any future development.

## Long explanation

To decide where to start contributing I extracted 3 metrics from the codebase:
1. Biggest files
2. Most complex files
3. Most changed files

#### 1. Biggest files

Using [cloc](https://github.com/AlDanial/cloc):
```
$ cloc --by-file doozerlib
-----------------------------------------------------------------------------------------
File                                       blank        comment           code
-----------------------------------------------------------------------------------------
doozerlib/distgit.py                         305            331           1149
doozerlib/runtime.py                         175            157            627
doozerlib/rpmcfg.py                           64             62            237
doozerlib/repos.py                            45             53            216
...
```

#### 2. Most complex files

Using [radon](https://pypi.org/project/radon/):
```
$ radon cc doozerlib -a -nc
doozerlib/distgit.py
    M 1122:4 ImageDistGitRepo.update_distgit_dir - F
    M 681:4 ImageDistGitRepo.build_container - D
    M 479:4 ImageDistGitRepo.push_image - D
    M 325:4 ImageDistGitRepo._generate_odcs_config - D
    M 1427:4 ImageDistGitRepo._merge_source - D
    M 870:4 ImageDistGitRepo._build_container - C
    M 999:4 ImageDistGitRepo._detect_permanent_build_failures - C
    C 271:0 ImageDistGitRepo - C
    M 1050:4 ImageDistGitRepo._mangle_yum - C
    M 1336:4 ImageDistGitRepo._update_csv - C
doozerlib/repos.py
    M 100:4 Repo.conf_section - C
doozerlib/image.py
    M 91:4 ImageMetadata.get_rpm_install_list - C
doozerlib/model.py
    M 105:4 ListModel._element_can_match - C
doozerlib/runtime.py
    M 193:4 Runtime.initialize - F
    M 654:4 Runtime.resolve_source - C

15 blocks (classes, functions, methods) analyzed.
Average complexity: D (21.8666666667)
```

#### 3. Most changed files

Only considering commits from last year, to show which files are _currently_ being frequently changed.
```
$ git log --pretty=format: --name-only --after="1 year ago" -- doozerlib | sort -n | uniq -c | sort -rg | head
    137 
     60 doozerlib/distgit.py
     53 doozerlib/VERSION
     27 doozerlib/runtime.py
     12 doozerlib/rpmcfg.py
      8 doozerlib/image.py
      8 doozerlib/distgit_test.py
      8 doozerlib/config.py
      7 doozerlib/schema_group.yml
      6 doozerlib/schema_image.yml
```

#### Conclusion

`doozerlib/distgit.py` was the "winner" in all 3 categories, so making it better will benefit us the most.

### CODE REVIEW

**Comment by thiagoalessio** [comment]
Oh! What a nice surprise: Travis already ran it the way I intended ;D
BTW, I'm specially curious to hear your opinions on the travis configuration ... it is not "the right way", but i wanted to have an env closer to ours (dev and prod).

File: tests/distgit_test.py [hunk 1]
```diff
@@ -0,0 +1,777 @@
+#!/usr/bin/env
```
Comment (shiywang) [inline tests/distgit_test.py]: Is this file just bump from `doozerlib/distgit_test.py` any reason not bumping the rest unittest files as well, like rpmcfg_test, runtime_test etc... ?

**Comment by shiywang** [comment]
Finally, we will have our own CI,  this is great work !! @thiagoalessio  : )

File: tests/distgit_test.py (hunk 1 above)
Comment (thiagoalessio) [reply tests/distgit_test.py]: yes, still a @wip, just started doing one by one (since some of the other *_test.py files contain failing tests)

**Comment by adammhaile** [comment]
@thiagoalessio 
`BTW, I'm specially curious to hear your opinions on the travis configuration ... it is not "the right way", but i wanted to have an env closer to ours (dev and prod).`

^^ What do you mean "not the right way"?

Also, what @shiywang said. I'm the lead on Doozer, so please always include me ;)

File: doozerlib/distgit.py [hunk 2]
```diff
@@ -262,6 +262,11 @@ def release_is_recent(self, release):
         )
         return False
 
+    # @TODO: check with my peers why this method was not implemented in the
+    # base class. Looks like the parent knows too much about the children.
```
Comment (adammhaile) [inline doozerlib/distgit.py]: Because the base class represents a dist-git, not an image and `_read_master_data` is poorly named (should probably be `_read_nvr_from_dockerfile` or similar. That function deals specifically with reading data out of the `Dockerfile` while distgits can also represent RPMs, which obviously wouldn't have a `Dockerfile`.

**Comment by thiagoalessio** [comment]
Hi @adammhaile,
by _not the right way_ I meant that I'm not leveraging Travis's environment, I create and run everything inside a container.

**Comment by adammhaile** [comment]
@thiagoalessio - I see absolutely nothing wrong with that. Travis is finicky and I've had my tests fail because they changed something in their base image. Running it as a docker image *should* isolate us from that. I guess that's why I was confused - I was impressed you did it the way you did :)

File: .ci/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+RUN yum install -y epel-release make gcc krb5-devel python-devel
+RUN yum install -y python-pip
+RUN pip install koji
+
+WORKDIR /code
+COPY requirements* ./
+RUN pip install -r requirements-test.txt
+
+ENV PYTHONPATH /code/doozerlib
```
Comment (sosiouxme) [inline .ci/Dockerfile]: could we also add:
```
ENV PYTHONDONTWRITEBYTECODE=1
```
since in a local test we're mounting in the code dir, don't want to write all the `.pyc` files.

File: tests/distgit/mocks.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    def detect_remote_source_branch(self, _):
+        pass
+
+class MockMetadata(object):
+
+    def __init__(self, runtime):
+        self.config = MockConfig()
+        self.runtime = runtime
+        self.logger = runtime.logger
+        self.name = "test"
```
Comment (sosiouxme) [inline tests/distgit/mocks.py]: I realize that this is just copied over from what we had, and I don't expect you to fix everything in the first pass. But just to set the expectation for how test infra should develop: if at all possible, no one should ever have to look at a mock definition to figure out what they're going to get back from a test. The mock should not fill in values like a fixture (or, if required for functioning at all, these should be explicitly dummy values), and the tests should set any values that they expect to see returned. That way you only have to look at the test method and the code it's testing to understand it.

That's how I summarize the advice in https://testing.googleblog.com/2018/02/testing-on-toilet-cleanly-create-test.html (that's about helper methods but these are essentially helper classes...)

```suggestion
        self.name = None
```

File: .ci/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,14 @@
+FROM centos:7
+
+RUN yum install -y epel-release make gcc krb5-devel python-devel
+RUN yum install -y python-pip
```
[the remark this answers is not in the document]
Comment (thiagoalessio) [reply .ci/Dockerfile]: Definitely! That's one of the topics I wanted to address. Do we have a dockerhub account that we could put that kind of stuff ?

File: .ci/Dockerfile (hunk 3 above)
Comment (thiagoalessio) [reply .ci/Dockerfile]: +1 adding now ;D

File: tests/distgit/mocks.py (hunk 4 above)
Comment (thiagoalessio) [reply tests/distgit/mocks.py]: :heavy_check_mark: Yes, I'm totally aligned with that idea! No hidden stuff, explicitly setting whatever is needed directly on the test function body is much cleaner.

File: .ci/Dockerfile (hunk 3 above)
Comment (thiagoalessio) [reply .ci/Dockerfile]: Resolved in commit 52330677

**Comment by thiagoalessio** [comment]
Closing this PR in favor of #101 #102 #103 #104 #105 #106 #107 #108 and many more to come ...
